invisible_cities/core/testing_utils_test.py

```diff
@@ -41,3 +42,98 @@ def test_assert_tables_equality_withNaN():
+
+    table2 = table1.copy()
+    table2[1][index] = value
+    with raises(AssertionError):
+        assert_tables_equality(table1, table2)
+
+
+def test_assert_tables_equality_fails_different_names():
+    # modify the type of a column and check that the function picks up
+    # the difference
+    dtypes1 = [('text', 'U10'), ('integer', 'i4'), ('float', 'f4')]
+    table1  = np.array([ ('one', 1, 1.0)
+                       , ('two', 2, 2.0)],
+                       dtype=dtypes1)
+
+    dtypes2 = [('different_name', 'U10'), ('integer', 'i4'), ('float', 'f4')]
+    table2  = np.array([ ('one', 1, 1.0)
+                       , ('two', 2, 2.0)],
+                       dtype=dtypes2)
+
+    with raises(AssertionError):
+        assert_tables_equality(table1, table2)
+
+
+def test_assert_tables_equality_fails_different_values():
+    # modify the type of a column and check that the function picks up
+    # the difference
+    dtypes1 = [('text', 'U10'), ('integer', 'i4'), ('float', 'f4')]
+    table1  = np.array([ ('one', 1, 1.0)
+                       , ('two', 2, 2.0)],
+                       dtype=dtypes1)
+
+    dtypes2    = list(dtypes1)
+    dtypes2[1] = ("integer", "f4")
+    table2     = table1.copy().astype(dtypes2)
+
+    with raises(AssertionError):
+        assert_tables_equality(table1, table2)
+
+
+@mark.parametrize("dtype", (int, float))
+def test_assert_tables_equality_equal_arrays(dtype):
+    array1 = np.arange(20, dtype=dtype)
+    array2 = array1.copy()
+    assert_tables_equality(array1, array2)
+
+
+@mark.parametrize("dtype", (int, float))
+def test_assert_tables_equality_different_arrays(dtype):
+    array1 = np.arange(20, dtype=dtype)
+    array2 = array1 + 1
+    with raises(AssertionError):
+        assert_tables_equality(array1, array2)
+
+
+@mark.parametrize(    "shape1  shape2".split()
+                 , ( ( (3, 4), (4, 4) )    # different lengths
+                   , ( (3, 3), (3, 4) )))  # different widths
+def test_assert_tables_equality_different_array_shapes(shape1, shape2):
+    array1 = np.ones(shape1)
+    array2 = np.ones(shape2)
+    with raises(AssertionError):
+        assert_tables_equality(array1, array2)
+
+
+def test_assert_tables_equality_different_table_shapes():
+    table1 = np.array([ (1, 1.0)
+                      , (2, 2.0)],
+                      dtype=[('integer', 'i4'), ('float', 'f4')])
+
+    table2 = np.array([ (1, 1.0)
+                      , (2, 2.0)
+                      , (3, 3.0)],
+                      dtype=[('integer', 'i4'), ('float', 'f4')])
+
+    table3 = np.array([ ('one', 1, 1.0)
+                      , ('two', 2, 2.0)],
+                      dtype=[('text', 'U10'), ('integer', 'i4'), ('float', 'f4')])
+
+    # different lengths
+    with raises(AssertionError):
+        assert_tables_equality(table1, table2)
+
+    # different widths
+    with raises(AssertionError):
+        assert_tables_equality(table1, table3)
```

Comment:
Does this need to be broken up into two tests still, now that there is only one condition for checking the shape?

Reply:
Do you mean broken into two different _assertions_? It does. Because execution is interrupted when the exception is raised and it resumes outside the `with raises` statement. i.e.
```
    with raises(AssertionError):
        assert_tables_equality(table1, table2)
        print("this line will not be executed if the one above raises an exception")
        
    print("but this one will")
```

Reply:
My fear is that if the lengths are the same (so the test fails) but the widths are different, the widths AssertionError won't be tested. Perhaps I'm misunderstanding the way the code works here?

Reply:
Let me see if I understand what you are saying. You are concerned that in the case where the test fails in the first scenario (different lengths), you don't know if the second scenario would also fail. Is this right?
If that is the case, it's not a big deal. Maybe having it in separate tests would give you a bit more information. In this case I feel like we don't gain enough by splitting the test in two, but if you feel otherwise I'll be happy to do so.

Reply:
No I think it's fine. In testing if the first case fails, fixing it and testing again would demonstrate whether the second case fails. I just wanted to make sure I wasn't misunderstanding the intent.